apps/encryption/lib/Services/PassphraseService.php

```diff
@@ -0,0 +1,142 @@
+<?php
+
+declare(strict_types=1);
+/**
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+namespace OCA\Encryption\Services;
+
+use OCA\Encryption\Crypto\Crypt;
+use OCA\Encryption\KeyManager;
+use OCA\Encryption\Recovery;
+use OCA\Encryption\Session;
+use OCA\Encryption\Util;
+use OCP\Encryption\Exceptions\GenericEncryptionException;
+use OCP\IUser;
+use OCP\IUserManager;
+use OCP\IUserSession;
+use Psr\Log\LoggerInterface;
+
+class PassphraseService {
+
+	/** @var array<string, bool> */
+	private static array $passwordResetUsers = [];
+
+	public function __construct(
+		private Util $util,
+		private Crypt $crypt,
+		private Session $session,
+		private Recovery $recovery,
+		private KeyManager $keyManager,
+		private LoggerInterface $logger,
+		private IUserManager $userManager,
+		private IUserSession $userSession,
+	) {
+	}
+
+	public function setProcessingReset(string $uid, bool $processing = true): void {
+		if ($processing) {
+			self::$passwordResetUsers[$uid] = true;
+		} else {
+			unset(self::$passwordResetUsers[$uid]);
+		}
+	}
+
+	/**
+	 * Change a user's encryption passphrase
+	 */
+	public function setPassphraseForUser(string $userId, string $password, ?string $recoveryPassword = null): bool {
+		// if we are in the process to resetting a user password, we have nothing
+		// to do here
+		if (isset(self::$passwordResetUsers[$userId])) {
+			return true;
+		}
+
+		// Check user exists on backend
+		$user = $this->userManager->get($userId);
+		if ($user === null) {
+			return false;
+		}
+
+		// Get existing decrypted private key
+		$currentUser = $this->userSession->getUser();
+
+		// current logged in user changes his own password
+		if ($currentUser !== null && $userId === $currentUser->getUID()) {
+			$privateKey = $this->session->getPrivateKey();
+
+			// Encrypt private key with new user pwd as passphrase
+			$encryptedPrivateKey = $this->crypt->encryptPrivateKey($privateKey, $password, $userId);
+
+			// Save private key
+			if ($encryptedPrivateKey !== false) {
+				$key = $this->crypt->generateHeader() . $encryptedPrivateKey;
+				$this->keyManager->setPrivateKey($userId, $key);
+				return true;
+			}
+
+			$this->logger->error('Encryption could not update users encryption password');
+			
+			// NOTE: Session does not need to be updated as the
+			// private key has not changed, only the passphrase
+			// used to decrypt it has changed
+		} else {
+			// admin changed the password for a different user, create new keys and re-encrypt file keys
+			$recoveryPassword = $recoveryPassword ?? '';
+			$this->initMountPoints($user);
+
+			$recoveryKeyId = $this->keyManager->getRecoveryKeyId();
+			$recoveryKey = $this->keyManager->getSystemPrivateKey($recoveryKeyId);
+			try {
+				$this->crypt->decryptPrivateKey($recoveryKey, $recoveryPassword);
+			} catch (\Throwable) {
```

Comment:
```suggestion
			} catch (\Exception) {
```
Avoid catch Throwable except for specific things (like, if you have a plugin system and call code from other apps, sure, catch Throwable, but when you call your own code, catch Exception and fix other errors.)